File: apps/frontend/src/types/preferences.ts
// FlowAudit User Preferences Types

export type ThemeMode = 'light' | 'dark' | 'system'
export type AccentColor = 'blue' | 'purple' | 'green' | 'orange' | 'pink' | 'teal'
export type ViewType = 'tiles' | 'list' | 'tree' | 'radial' | 'minimal'
export type FontSize = 'small' | 'medium' | 'large'
export type TimeFormat = '12h' | '24h'
export type EmailDigest = 'none' | 'daily' | 'weekly'

export type WidgetId =
  | 'active_audits'
  | 'appointments'
  | 'tasks'
  | 'activities'
  | 'statistics'
  | 'quick_actions'

export interface AppearancePreferences {
  theme: ThemeMode
  accentColor: AccentColor
  fontSize: FontSize
  reducedMotion: boolean
  highContrast: boolean
}

export interface NavigationPreferences {
  defaultView: ViewType
  sidebarCollapsed: boolean
  enableAnimations: boolean
  enableHoverEffects: boolean
  showModuleBadges: boolean
}

export interface DashboardPreferences {
  startPage: string
  visibleWidgets: WidgetId[]
  widgetOrder: WidgetId[]
  compactMode: boolean
}

export interface ModulePreference {
  defaultTab?: string
  sortOrder?: string
  filters?: Record<string, unknown>
  columns?: string[]
}

export interface NotificationPreferences {
  enabled: boolean
  sound: boolean
  desktop: boolean
  emailDigest: EmailDigest
}

export interface LocalePreferences {
  language: 'de' | 'en'
  dateFormat: string
  timeFormat: TimeFormat
  timezone: string
  numberFormat: string
}

export interface ShortcutPreferences {
  enabled: boolean
  custom: Record<string, string>
}

export interface UserPreferences {
  id?: string
  userId?: string

  appearance: AppearancePreferences
  navigation: NavigationPreferences
  dashboard: DashboardPreferences
  modulePreferences: Record<string, ModulePreference>
  notifications: NotificationPreferences
  locale: LocalePreferences
  shortcuts: ShortcutPreferences

  updatedAt?: string
}

export const defaultPreferences: UserPreferences = {
  appearance: {
    theme: 'system',
    accentColor: 'blue',
    fontSize: 'medium',
    reducedMotion: false,
    highContrast: false
  },
  navigation: {
    defaultView: 'tiles',
    sidebarCollapsed: false,
    enableAnimations: true,
    enableHoverEffects: true,
    showModuleBadges: true
  },
  dashboard: {
    startPage: '/dashboard',
    visibleWidgets: ['active_audits', 'appointments', 'tasks', 'activities'],
    widgetOrder: ['active_audits', 'appointments', 'tasks', 'activities'],
    compactMode: false
  },
  modulePreferences: {},
  notifications: {
    enabled: true,
    sound: false,
    desktop: true,
    emailDigest: 'daily'
  },
  locale: {
    language: 'de',
    dateFormat: 'DD.MM.YYYY',
    timeFormat: '24h',
    timezone: 'Europe/Berlin',
    numberFormat: 'de-DE'
  },
  shortcuts: {
    enabled: true,
    custom: {}
  }
}

export const accentColors: Record<AccentColor, { name: string; hex: string }> = {
  blue: { name: 'Ocean', hex: '#3b82f6' },
  purple: { name: 'Lavender', hex: '#a855f7' },
  green: { name: 'Forest', hex: '#10b981' },
  orange: { name: 'Sunset', hex: '#f97316' },
  pink: { name: 'Rose', hex: '#ec4899' },
  teal: { name: 'Teal', hex: '#14b8a6' }
}

export const viewTypes: Record<ViewType, { name: string; icon: string; description: string }> = {
  tiles: {
    name: 'Kacheln',
    icon: 'grid',
    description: 'Übersichtliche Karten mit Icons und Statistiken'
  },
  list: {
    name: 'Liste',
    icon: 'list',
    description: 'Platzsparend, ideal für Power-User'
  },
  tree: {
    name: 'Sidebar',
    icon: 'sidebar',
    description: 'Klassische Navigation mit Hierarchie'
  },
  radial: {
    name: 'Radial',
    icon: 'circle',
    description: 'Visuell beeindruckend, zentrale Nutzerposition'
  },
  minimal: {
    name: 'Minimal',
    icon: 'minus',
    description: 'Maximale Aufgeräumtheit, nur das Wesentliche'
  }
}

File: apps/backend/app/schemas/preferences.py
"""Preferences Pydantic Schemas"""

from typing import Optional, Literal
from pydantic import BaseModel, Field


class AppearancePreferences(BaseModel):
    theme: Literal["light", "dark", "system"] = "system"
    accentColor: Literal["blue", "purple", "green", "orange", "pink", "teal"] = "blue"
    fontSize: Literal["small", "medium", "large"] = "medium"
    reducedMotion: bool = False
    highContrast: bool = False


class NavigationPreferences(BaseModel):
    defaultView: Literal["tiles", "list", "tree", "radial", "minimal"] = "tiles"
    sidebarCollapsed: bool = False
    enableAnimations: bool = True
    enableHoverEffects: bool = True
    showModuleBadges: bool = True


class DashboardPreferences(BaseModel):
    startPage: str = "/dashboard"
    visibleWidgets: list[str] = Field(
        default_factory=lambda: ["active_audits", "appointments", "tasks", "activities"]
    )
    widgetOrder: list[str] = Field(
        default_factory=lambda: ["active_audits", "appointments", "tasks", "activities"]
    )
    compactMode: bool = False


class NotificationPreferences(BaseModel):
    enabled: bool = True
    sound: bool = False
    desktop: bool = True
    emailDigest: Literal["none", "daily", "weekly"] = "daily"


class LocalePreferences(BaseModel):
    language: Literal["de", "en"] = "de"
    dateFormat: str = "DD.MM.YYYY"
    timeFormat: Literal["12h", "24h"] = "24h"
    timezone: str = "Europe/Berlin"
    numberFormat: str = "de-DE"


class ShortcutPreferences(BaseModel):
    enabled: bool = True
    custom: dict[str, str] = Field(default_factory=dict)


class UserPreferencesSchema(BaseModel):
    id: Optional[str] = None
    userId: Optional[str] = None

    appearance: AppearancePreferences = Field(default_factory=AppearancePreferences)
    navigation: NavigationPreferences = Field(default_factory=NavigationPreferences)
    dashboard: DashboardPreferences = Field(default_factory=DashboardPreferences)
    modulePreferences: dict = Field(default_factory=dict)
    notifications: NotificationPreferences = Field(
        default_factory=NotificationPreferences
    )
    locale: LocalePreferences = Field(default_factory=LocalePreferences)
    shortcuts: ShortcutPreferences = Field(default_factory=ShortcutPreferences)

    updatedAt: Optional[str] = None

    class Config:
        from_attributes = True


class UserPreferencesUpdate(BaseModel):
    """Schema for updating preferences - all fields optional"""

    appearance: Optional[AppearancePreferences] = None
    navigation: Optional[NavigationPreferences] = None
    dashboard: Optional[DashboardPreferences] = None
    modulePreferences: Optional[dict] = None
    notifications: Optional[NotificationPreferences] = None
    locale: Optional[LocalePreferences] = None
    shortcuts: Optional[ShortcutPreferences] = None
